File: examples/web-programs/response_diff.mjs
import {open} from "node:fs/promises";
import {createHash} from "node:crypto";
async function load(path) {
  const file=await open(path,"r"), buffer=Buffer.alloc(1048577);let used=0;
  try {while(used<buffer.length) {const {bytesRead}=await file.read(buffer,used,buffer.length-used,null);if(!bytesRead)break;used+=bytesRead;}}
  finally {await file.close();}
  if(used>1048576)throw new Error("File too large");
  const row=JSON.parse(new TextDecoder("utf-8",{fatal:true}).decode(buffer.subarray(0,used)));
  if(!row || Array.isArray(row) || !Number.isInteger(row.status) || row.status<100 || row.status>599 || typeof row.body!=="string")throw new Error("Invalid response shape");
  return row;
}
try {
  if(process.argv.length!==4)throw new Error("Usage: node response_diff.mjs baseline.json comparison.json");
  const a=await load(process.argv[2]),b=await load(process.argv[3]);
  console.log(JSON.stringify({same_status:a.status===b.status,same_body:a.body===b.body,body_bytes:[a,b].map(x=>Buffer.byteLength(x.body,"utf8")),sha256:[a,b].map(x=>createHash("sha256").update(x.body,"utf8").digest("hex")),note:"Different responses do not prove broken authorization. Verify identity, expected permissions and returned data."}));
} catch {console.error("Cannot compare: invalid or unreadable input");process.exitCode=1;}

File: examples/web-programs/response_diff.py
import hashlib, json, sys
from pathlib import Path

def load(path):
    with Path(path).open("rb") as f: data = f.read(1048577)
    if len(data)>1048576: raise ValueError("File too large")
    row=json.loads(data.decode("utf-8"))
    if not isinstance(row,dict) or type(row.get("status")) is not int or not 100<=row["status"]<=599 or not isinstance(row.get("body"),str):
        raise ValueError("Expected status integer 100..599 and body string")
    return row

def compare(a,b):
    return {"same_status":a["status"]==b["status"], "same_body":a["body"]==b["body"], "body_bytes":[len(x["body"].encode("utf-8")) for x in (a,b)], "sha256":[hashlib.sha256(x["body"].encode("utf-8")).hexdigest() for x in (a,b)], "note":"Different responses do not prove broken authorization. Verify identity, expected permissions and returned data."}

if __name__=="__main__":
    if len(sys.argv)!=3: sys.exit("Usage: python3 response_diff.py baseline.json comparison.json")
    try: print(json.dumps(compare(load(sys.argv[1]),load(sys.argv[2])),sort_keys=True))
    except (ValueError,OSError): sys.exit("Cannot compare: invalid or unreadable input")
